src/mpi/request/testall.c

```diff
@@ -35,148 +35,34 @@ int MPI_Testall(int count, MPI_Request array_of_requests[], int *flag,
+        if (request_ptrs[i] != NULL) {
+            if (request_ptrs[i]->kind == MPIR_REQUEST_KIND__GREQUEST &&
+                    request_ptrs[i]->u.ureq.greq_fns->poll_fn != NULL)
             {
-                mpi_errno = MPI_ERR_IN_STATUS;
-                MPIR_ERR_SET(rc, MPIX_ERR_PROC_FAILED_PENDING, "**failure_pending");
-                status_ptr = (array_of_statuses != MPI_STATUSES_IGNORE) ? &array_of_statuses[i] : MPI_STATUS_IGNORE;
-                if (status_ptr != MPI_STATUS_IGNORE) status_ptr->MPI_ERROR = rc;
-                proc_failure = TRUE;
+                mpi_errno = (request_ptrs[i]->u.ureq.greq_fns->poll_fn)(request_ptrs[i]->u.ureq.greq_fns->grequest_extra_state,
+                        &(array_of_statuses[i]));
+                if (mpi_errno != MPI_SUCCESS) goto fn_fail;
             }
-        }
-    }
-
-    if (n_completed == count || mpi_errno == MPI_ERR_IN_STATUS)
-    {
-        n_completed = 0;
-        for (i = 0; i < count; i++)
-        {
-            if (request_ptrs[i] != NULL)
-            {
-                if (MPIR_Request_is_complete(request_ptrs[i]))
-                {
-                    n_completed ++;
-                    status_ptr = (array_of_statuses != MPI_STATUSES_IGNORE) ? &array_of_statuses[i] : MPI_STATUS_IGNORE;
-                    rc = MPIR_Request_complete(&array_of_requests[i], request_ptrs[i], status_ptr, &active_flag);
-                    if (mpi_errno == MPI_ERR_IN_STATUS && status_ptr != MPI_STATUS_IGNORE)
-                    {
-                        if (active_flag)
-                        {
-                            status_ptr->MPI_ERROR = rc;
-                        }
-                        else
-                        {
-                            status_ptr->MPI_ERROR = MPI_SUCCESS;
-                        }
-                    }
-                }
-                else
-                {
-                    if (mpi_errno == MPI_ERR_IN_STATUS && array_of_statuses != MPI_STATUSES_IGNORE)
-                    {
-                        if (!proc_failure)
-                            array_of_statuses[i].MPI_ERROR = MPI_ERR_PENDING;
-                        else
-                            array_of_statuses[i].MPI_ERROR = MPIX_ERR_PROC_FAILED_PENDING;
-                    }
-                }
-            }
-            else
-            {
-                n_completed ++;
-                if (array_of_statuses != MPI_STATUSES_IGNORE)
-                {
-                    MPIR_Status_set_empty(&array_of_statuses[i]);
-                    if (mpi_errno == MPI_ERR_IN_STATUS)
-                    {
-                        array_of_statuses[i].MPI_ERROR = MPI_SUCCESS;
-                    }
-                }
+            if (!MPIR_Request_is_complete(request_ptrs[i])) {
+                goto fn_exit;
             }
         }
     }
-
-    *flag = (n_completed == count) ? TRUE : FALSE;
+    *flag = TRUE;
```

Comment:
`flag=TRUE` implies all requested have completed. This commit breaks this semantic by dropping `n_completed` and unconditionally setting `flag`

Reply:
Not sure if I understand this, if there is an uncompleted request It should have return and never arrive here.

Reply:
right, never mind. I thought I was looking at testany/some

Reply:
but `n_complete` is still unused. Can you compile locally with `--enable-strict` warnings of this type?

Reply:
Yes, I'm cleaning those up. Thanks for the useful flag!